clang/include/clang/CIR/Dialect/IR/CIRTypes.td

```diff
@@ -111,7 +112,7 @@ def CIR_StructType : CIR_Type<"Struct", "struct",
     "bool":$body,
     "bool":$packed,
     "mlir::cir::StructType::RecordKind":$kind,
-    "std::optional<::mlir::cir::ASTRecordDeclAttr>":$ast
+    OptionalParameter<"Attribute">:$ast
```

Comment:
Can we keep `ASTRecordDeclAttr` here so that it's strong-typed?

Reply:
Nope, because this locks this to use only `ASTRecordDeclAttr`. The point of this PR is to allow to use CIR dialect without need to have AST in hand. We plan to lower [vast](https://github.com/trailofbits/vast) dialects and use its high-level representation to provide AST API.

For type checking we can probably put corresponding interface here, but it disallows upcasts, which are used on the attributes later in code (e.g. `dyn_cast` to `ClassTemplateSpecializationDecl`)

Reply:
I'm gonna give this a bit more thought, but if this is the only possible solution, then `CIR_StructType` verifier needs to make sure that (a) when it has some attached AST, it (b) has to be one of the possibilities we know about.

Reply:
I was also playing around with `ASTRecordLikeAttr` which is defined as `AnyOf< ASTRecordDeclInterface, ASTCXXRecordInterface, ... >` which lists all `AST` posibilities.

Reply:
Sounds like an `ASTFunctionLikeAttr` would also be helpful!

Reply:
fixed in 7571992233389ba6067eb5ffcceacea169e23144